src/analyses/useAfterFree.ml

```diff
@@ -125,9 +149,25 @@ struct
     | StartOf lval
     | AddrOf lval -> warn_lval_might_contain_freed ~is_double_free transfer_fn_name ctx lval
 
-  let side_effect_mem_free ctx freed_heap_vars threadid =
-    let threadid = G.singleton threadid in
-    D.iter (fun var -> ctx.sideg var threadid) freed_heap_vars
+  let side_effect_mem_free ctx freed_heap_vars threadid joined_threads =
+    let side_effect_globals_to_heap_var heap_var globals_to_add =
+      let (tid_to_add, joined_threads_to_add) = globals_to_add in
+      let current_globals = ctx.global heap_var in
+      (*
+        * Check if there are tuples with the same first component (i.e., tid)
+        * If no, side-effect the globals that we receive here and be done
+        * If yes, join all second components together and side-effect a single global with
+        the tid as first component and the joined second components as a single second component
+      *)
+      let globals_with_same_tid = G.filter (fun (tid, _) -> ThreadIdDomain.ThreadLifted.equal tid tid_to_add) current_globals in
```

Comment:
This sounds like `G` should be a `MapDomain`, not a `SetDomain` of pairs where a complete iteration is done to check equality on first component. Or am I mistaken?